beamline.rst

```diff
@@ -1,212 +1,379 @@
+        display_name: the name to display in the user interface i.e. "Quick Realign"
+        tool_tip: the tool tip
+        messages: any initial messages to display
+        """
+
+        name: str   # NB This assumes that only one instance of a command can run at any time
+        state: CmdState
+        args: tuple[str, ...]
+        kwargs: dict[str, Any]
+        display_name: Optional[str] # defaults to name
+        tool_tip: Optional[str]
+        messages: Tuple[str]
+
 
+    def get_procedures() -> OrderedDict[str,ProcedureData]:
+        """
+        :returns: an OrderedDict name:ProcedureData, each describing a procedure,
+                  such as: realignment, anneal
 
-    def get_procedures() -> List[ProcedureData]:
-    """
-    :returns: a list of tuples each describing a beamline procedure,
-              such as: realignment, anneal
+        # NB we may want nested dictionaries if procedures are divided by component
 
-    :rtype: List[ProcedureData]
-    """
+        :rtype: Tuple[ProcedureData]
+        """
         pass
 
-        
-    def run_procedure(cmd_id:str, args) -> bool:
-    """
-    Runs procedure identified by cmd_id passes args as its arguments
-    
-    (Errors and progress that occurs is passed asynchronously via the
-     available signaling mechanism.)
-    """
-        pass    
-    
-    def stop_procedure(cmd_id:str) -> bool:
-    """
-    Stops a running procedure identified by cmd_id
-
-    (Signal emitted if stopped or on timeout waiting for procedure to stop)
-    """
+
+    def run_procedure(name:str, *args, **kwargs) -> bool:
+        """
+        Runs procedure identified by name with mandatory arguments args
+        and optional keyword arguments kwargs
+
+        (Errors and progress that occurs is passed asynchronously via the
+         available signalling mechanism.)
+
+        # NB progress handling is still unresolved!
+
+        """
+        pass
+
+    def stop_procedure(name:str) -> bool:
+        """
+        Stops a running procedure identified by name
+
+        (Signal emitted if stopped or on timeout waiting for procedure to stop)
+
+        # NB we are assuming that names (like 'Quick_Realign') are unique,
+        #    and that a given procedure cannot run in two parallel processes.
+        """
         pass
 
-    class ActuatorData(NamedTuple):
-    """
-    Describes an actuator commonly a motor, that is considered to be
-    "globally" available on the beamline. Such actuators could be shutters
-    of different kinds, i.e: fast shutter, safety shutter, other in-out
-    devices; beam stop, detector cover or even motors such as omega.
+    class CmdState(Enum):
+        # NB do we need other / different states here?
+        UNUSABLE = 0
+        READY = 1
+        RUNNING = 2
+
+  **# ACTUATORS**
+
+    class GenericActuatorData(NamedTuple):
+        """
+        Describes a generic actuator, in practice any attribute that could be
```

Comment:
Yes, it should be a well-defined subset (see below for a proposal). But as you will see I think it would work both for Kappa, resolution, beam_shape, beamstop, and wavelength, ... and that we should use the same mechanism even on fixed-energy beamlines (just setting the state to 'FIXED'). Then I can get hole of the wavelength in the same way without first having to check if the beamline is tunable.